mkosi/__init__.py

```diff
@@ -6623,33 +6634,47 @@ def interface_exists(dev: str) -> bool:
 
 def find_address(args: CommandLineArguments) -> Tuple[str, str]:
     name = virt_name(args)
+    timeout = float(args.ssh_timeout)
 
-    if interface_exists(f"ve-{name}"):
-        dev = f"ve-{name}"
-    elif interface_exists(f"vt-{name}"):
-        dev = f"vt-{name}"
-    else:
-        die("Container/VM interface not found")
+    while timeout >= 0:
+        last_exception = None
+        stime = time.time()
+        try:
+            if interface_exists(f"ve-{name}"):
+                dev = f"ve-{name}"
+            elif interface_exists(f"vt-{name}"):
+                dev = f"vt-{name}"
+            else:
+                raise MkosiException("Container/VM interface not found")
 
-    link = json.loads(run(["ip", "-j", "link", "show", "dev", dev], stdout=PIPE, text=True).stdout)[0]
-    if link["operstate"] == "DOWN":
-        die(f"{dev} is not enabled. Make sure systemd-networkd is running so it can manage the interface.")
+            link = json.loads(run(["ip", "-j", "link", "show", "dev", dev], stdout=PIPE, text=True).stdout)[0]
+            if link["operstate"] == "DOWN":
+                raise MkosiException(
+                    f"{dev} is not enabled. Make sure systemd-networkd is running so it can manage the interface."
+                )
 
-    # Trigger IPv6 neighbor discovery of which we can access the results via `ip neighbor`. This allows us to
-    # find out the link-local IPv6 address of the container/VM via which we can connect to it.
-    run(["ping", "-c", "1", "-w", "15", f"ff02::1%{dev}"], stdout=DEVNULL)
+            # Trigger IPv6 neighbor discovery of which we can access the results via `ip neighbor`. This allows us to
+            # find out the link-local IPv6 address of the container/VM via which we can connect to it.
+            run(["ping", "-c", "1", "-w", "15", f"ff02::1%{dev}"], stdout=DEVNULL)
+
+            for _ in range(50):
+                neighbors = json.loads(
+                    run(["ip", "-j", "neighbor", "show", "dev", dev], stdout=PIPE, text=True).stdout
+                )
 
-    for _ in range(50):
-        neighbors = json.loads(run(["ip", "-j", "neighbor", "show", "dev", dev], stdout=PIPE, text=True).stdout)
+                for neighbor in neighbors:
+                    dst = cast(str, neighbor["dst"])
+                    if dst.startswith("fe80"):
+                        return dev, dst
 
-        for neighbor in neighbors:
-            dst = cast(str, neighbor["dst"])
-            if dst.startswith("fe80"):
-                return dev, dst
+                time.sleep(0.4)
+        except MkosiException as e:
+            last_exception = str(e)
```

Comment:
I think we might not need the last_exception variable if you check if the timeout has expired in the except clause here and reraise the exception if the timeout has expired.

This also means that we'll always try to run through the loop once, even if the timeout is 0 (which I think is a good thing)

Reply:
That actually was the original approach (sans some minor details), but I didn't want to make the timeout calculation twice. But I guess it would look nicer than the current `last_exception` approach. I'll prep a fix shortly.

Reply:
Ah, simple `raise` would hide the actual error message, so we have to use either `die()` or `warn(); raise`, which is not as pretty.